docs/source/man/config.rst
.. _configuring:


Configuration files
*******************

All the configuration files are store in ``/etc/ajenti`` :

  * **config.yml**: the main configuration file with all important parameters,
  * **smtp.yml**: credentials for an email server relay, used for reset password functionality,
  * **users.yml**: the default file which contains user account for the user authentication provider.

All configuration files use the `yaml format <https://en.wikipedia.org/wiki/YAML>`_

config.yml in details
=====================

Ajenti will use the following parameters :

auth block
----------

.. code-block:: yaml

    auth:
      allow_sudo: true
      emails: {}
      provider: os
      users_file: /etc/ajenti/users.yml

Explanations:

  * **allow_sudo**: **true** or **false** (allow users in the sudo group to elevate)
  * **emails**: {} (not currently used)
  * **provider**: authentication method to use, **os** (users from the os) or **users**
  * **users_file**: if the users authentication provider is used, path to the users file (default **/etc/ajenti/users.yml**)

The parameter **user_config** was used to specified where the user configuration was stored, but is now deprecated, since it's bound to the **provider** (**os** or **users**) to avoid duplicates entries.

bind block
----------

.. code-block:: yaml

    bind:
      host: 0.0.0.0
      mode: tcp
      port: 8000

Explanations:

  * **host**: ip on which to listen (default **0.0.0.0**)
  * **mode**: type of socket, **tcp** or **unix**
  * **port**: port on which to listen, default **8000**

ssl block
---------

.. code-block:: yaml

    ssl:
      enable: true
      certificate: /etc/ajenti/mycert.pem
      fqdn_certificate: /etc/letsencrypt/ajenti.pem
      force: false
      client_auth:
         enable: true
         force: true
         certificates:
           digest: 15:E8:5E:E5:D2:E8:75:0D:53:FF:22:A8:79:28:E5:BE:33:E0:37:07:FB:31:47:4D:61:69:AB:43:F8:5B:23:78
           name: C=NA,ST=NA,O=sajenti.mydomain.com,CN=root@ajenti.mydomain.com
           serial: 352674123960898230347891590646542168839110009016
           user: root

Explanations:

  * **enable**: **true** or **false** to provide support for https. It's highly recommended to set it to **true**
  * **certificate**: full path to default global certificate, used to generate client certificates, and fot the https protocol, if the parameter ``fqdn_certificate`` is not set. The PEM file should contains the certificate itself, and the private key.
  * **fqdn_certificate**: full path certificate for your FQDN (e.g. ``/etc/ajenti/mycert.pem``). The PEM file should contains the certificate itself, and the private key.
  * **force**: spawn a small listener on port 80 to enable a redirect from ``http://hostname`` to ``https://hostname:port``.
  * **client_auth**:
     * **enable**: **true** or **false** to enable client authentication via certificates
     * **force**: if **true**, only allows login with client certificate. If **false**, also permit authentication with password
     * **certificates**: this entry contains all client certifcates for an automatic login. It will be filled through the settings in Ajenti with the following structure:
        * **digest**: digest of the certificate
        * **name**: name of the certificate
        * **serial**: serial of the certificate
        * **user**: username

email block
-----------

.. code-block:: yaml

    email:
      enable: true
      templates:
        reset_email : /etc/ajenti/email/mytemplate_for_reset_password.html

Explanations:

  * **enable**: **true** or **false**, if you want to enable the password reset function. But for this you need to set the smtp credentials in ``/etc/ajenti/smtp.yml``
  * **templates**:
    * **reset_email**: full path to template email for reset password functionality

The default template used to reset email password is located `here <https://github.com/ajenti/ajenti/blob/master/ajenti-core/aj/static/emails/reset_email.html>`_.
The variables are automatically filled with jinja2.

Other global parameters
-----------------------

.. code-block:: yaml

    color: blue
    language: en
    logo: /srv/dev/ajenti/ajenti-panel/aj/static/images/Logo.png
    max_sessions: 10
    name: ajenti.mydomain.com
    restricted_user: nobody
    session_max_time: 1200

Explanations:

  * **color**: secundary color of the CSS theme (possibles values are **default**, **bluegrey**, **red**, **deeporange**, **orange**, **green**, **teal**, **blue** and **purple**)
  * **language**: language prefence for all users, default **en**
  * **logo**: full path to your own logo, default is `the one from Ajenti <https://github.com/ajenti/ajenti/blob/master/ajenti-core/aj/static/images/Logo.png>`_
  * **max_sessions**: max number of simultaneously sessions, default is **99**. If the max is reached, the older inactive session will be deactivated
  * **name**: your domain name
  * **restricted_user**: user to use for the restricted functionalities, like for the login page. It's an important security parameter in order to limit the actions in restricted environments : all actions in restricted environments will be done with this user's privileges.
    Default is **nobody**.
  * **session_max_time**: max validity time in seconds before automatic logout.
    Default is **3600** (one hour).
  * **trusted_domains** ( `Ajenti` >= 2.2.1 ) : comma separated list of trusted domains under which it's possible to reach your `Ajenti` server. When the HTTP headers are tested, a valid origin will be considered as one of the domains listed. It's necessary to specify the protocol and the port. It's mean that an entry should look like `http://my.domain.com:8000`. If set, the first entry of this list will be used as url for the password reset functionality, for more security.
  * **trusted_proxies** ( `Ajenti` >= 2.2.1 ) : comma separated list of trusted proxies. This is actually used in order to get the real ip of the client.

smtp.yml in details
===================

This file contains all the credentials of an email server which can be used as email relay to send an email to reset a forgotten password. A mail backend is prepared in ajenti-core, and this could be possibly be used in the future to send some other notifications, but it's currently only used in order to reset a password.

.. code-block:: yaml

    smtp:
      password: MyVeryStrongStrongPassword
      port: starttls
      server: mail.mydomain.com
      user: mail@mydomain.com

Explanations:

     * **port**: **starttls** (will use 587) or **ssl** (will use 465)
     * **server**: server hostname, like ``mail.mydomain.com``
     * **user**: user to authenticate
     * **password**: password of the mail user

users.yml in details
====================

Ajenti gives the possibility to use two authentication methods : **os** or **users**. If **users** is used, all user informations are stored in **users_file**. It's automatically filled with the user plugin.


The default path for the **users_file** is ``/etc/ajenti/users.yml`` with following structure:

.. code-block:: yaml

    users:
      arnaud:
        email: arnaud@mydomain.com
        fs_root: /home/arnaud
        password: 73637279707.....
        permissions:
          packages:install: false
          sidebar:view:/view/cron: false
        uid: 1002

Explanations:

  * **password**: hash of the password (see :ref:`how_to_generate_password_hash`)
  * **permissions**: list of permissions of the user
  * **uid**: related os uid to run the worker on
  * **fs_root**: root directory
  * **email**: email to use for password reset.

.. _how_to_generate_password_hash:

how to generate password hash
=============================
Ajenti uses ``scrypt`` for encryption and hashing.

To "hash" your desired password with scrypt, you can do the following:

.. code-block:: py

  import scrypt, os
  
  bytes = scrypt.encrypt(os.urandom(256), "password", maxtime=1)
  hex = bytes.hex()
  print(hex)

Note that ``scrypt.encrypt`` doesn't produce a traditional hash. It produces a derived key.


docs/source/ref/aj.config.rst
API: aj.config
**************

.. automodule:: aj.config
   :members:
   :undoc-members:


CHANGELOG
ajenti (0.6.3) UNRELEASED; urgency=low

  * Bugfix release

 -- Eugeny Pankov <e@ajenti.org>  Tue, 4 Sep 2012 16:26:00 +0300

ajenti (0.6.2) UNRELEASED; urgency=low

  * Bugfix release

 -- Eugeny Pankov <e@ajenti.org>  Sat, 14 Apr 2011 13:38:00 +0300

ajenti (0.6.1) UNRELEASED; urgency=low

  * Bugfix release

 -- Eugeny Pankov <e@ajenti.org>  Thu, 31 Oct 2011 15:48:00 +0300

ajenti (0.6.0) UNRELEASED; urgency=low

  * Minor fixes

 -- Eugeny Pankov <e@ajenti.org>  Thu, 14 Oct 2011 23:16:00 +0300

ajenti (0.5.101) UNRELEASED; urgency=low

  * Hotfix for gevent crash on Debian Squeeze.

 -- Eugeny Pankov <e@ajenti.org>  Thu, 14 Oct 2011 23:16:00 +0300

ajenti (0.5.100) UNRELEASED; urgency=low

  * Hotfix for JS performance issue (Twipsy JS).

 -- Eugeny Pankov <e@ajenti.org>  Thu, 13 Oct 2011 22:41:00 +0300

ajenti (0.5.99) UNRELEASED; urgency=low

  * Initial pre-release.

 -- Eugeny Pankov <e@ajenti.org>  Thu, 13 Oct 2011 18:29:00 +0300


ajenti-core/aj/config.py
import os
import pwd
import stat
import yaml
import logging
from jadi import interface, component, service
import aj
from aj.util import public


class BaseConfig():
    """
    A base class for config implementations. Your implementation must be able to save
    arbitrary mixture of ``dict``, ``list``, and scalar values.

    .. py:attribute:: data

        currently loaded config content

    """
    def __init__(self):
        self.data = None

    def load(self):
        """
        Should load config content into :attr:`data`.
        """
        raise NotImplementedError()

    def save(self):
        """
        Should save config content from :attr:`data`.
        """
        raise NotImplementedError()

    def ensure_structure(self):
        # Global options
        self.data.setdefault('name', None)
        self.data.setdefault('trusted_domains', [])
        self.data.setdefault('trusted_proxies', [])
        self.data.setdefault('max_sessions', 99)
        self.data.setdefault('session_max_time', 3600)
        self.data.setdefault('language', 'en')
        self.data.setdefault('restricted_user', 'nobody')
        self.data.setdefault('logo', os.path.dirname(__file__) + '/static/images/Logo.png')

        # Main view
        self.data.setdefault('view', {})
        self.data['view'].setdefault('plugin', 'core')
        self.data['view'].setdefault('filepath', 'content/pages/index.html')

        # Authentication
        self.data.setdefault('auth', {})
        self.data['auth'].setdefault('emails', {})
        self.data['auth'].setdefault('provider', 'os')
        self.data['auth'].setdefault('users_file', '/etc/ajenti/users.yml')

        # SSL
        self.data.setdefault('ssl', {})
        self.data['ssl'].setdefault('enable', False)
        self.data['ssl'].setdefault('certificate', None)
        self.data['ssl'].setdefault('fqdn_certificate', None)
        self.data['ssl'].setdefault('force', False)
        self.data['ssl'].setdefault('client_auth', {})
        self.data['ssl']['client_auth'].setdefault('enable', False)
        self.data['ssl']['client_auth'].setdefault('force', False)
        self.data['ssl']['client_auth'].setdefault('certificates', [])
        if self.data['ssl']['client_auth']['certificates'] is None:
            self.data['ssl']['client_auth']['certificates'] = []

        # Emails
        self.data.setdefault('email', {})
        self.data['email'].setdefault('enable', False)
        self.data['email'].setdefault('templates', {})

        # Before Ajenti 2.1.38, the users were stored in config.yml
        if 'users' in self.data['auth'].keys():
            logging.warning(f"Users should be stored in {self.data['auth']['users_file']}, migrating it ...")
            self.migrate_users_to_own_configfile()

    def migrate_users_to_own_configfile(self):
        users_path = self.data['auth']['users_file']

        if os.path.isfile(users_path):
            logging.info(f"{users_path} already existing, backing it up")
            os.rename(users_path, users_path + '.bak')

        to_write = {'users': self.data['auth']['users']}
        with open(users_path, 'w') as f:
           f.write(yaml.safe_dump(to_write, default_flow_style=False, encoding='utf-8', allow_unicode=True).decode('utf-8'))

        del self.data['auth']['users']
        self.save()
        logging.info(f"{users_path} correctly written")


    def get_non_sensitive_data(self):
        return {
            'color': self.data['color'],
            'language': self.data['language'],
            'name': self.data['name'],
            'session_max_time': self.data['session_max_time'],

        }

class SmtpConfig(BaseConfig):
    """
    Class to handle the smtp config file in order to store credentials of the email
    server relay.
    Config file is located at /etc/ajenti/smtp.yml and should have the following
    structure :
    smtp:
      port: starttls or ssl
      server: myserver.domain.com
      user: user to authenticate
      password: password of the mail user
    """

    def __init__(self):
        BaseConfig.__init__(self)
        self.data = {}
        self.path = '/etc/ajenti/smtp.yml'

    def ensure_structure(self):
        self.data.setdefault('smtp', {})
        self.data['smtp'].setdefault('password', None)
        self.data['smtp'].setdefault('port', None)
        self.data['smtp'].setdefault('server', None)
        self.data['smtp'].setdefault('user', None)

    def get_smtp_password(self):
        # if smtp.yml is not provided
        if self.data['smtp']['password'] is None:
            return ''
        with open(self.path, 'r') as smtp:
            smtp_config = yaml.load(smtp, Loader=yaml.SafeLoader).get('smtp', {})
        return smtp_config.get('password', None)

    def load(self):
        if not os.path.exists(self.path):
            logging.error(f'Smtp credentials file "{self.path}" not found')
        else:
            if os.geteuid() == 0:
                os.chmod(self.path, 384)  # 0o600
                with open(self.path, 'r') as smtp:
                    self.data = yaml.load(smtp, Loader=yaml.SafeLoader) or {}
                    # Prevent password leak
                    self.ensure_structure()
                    self.data['smtp']['password'] = ''

    def save(self, data):
        # Prevent emptying password from settings plugin
        if not data['smtp']['password']:
            data['smtp']['password'] = self.get_smtp_password()
        with open(self.path, 'w') as smtp:
            smtp.write(
                yaml.safe_dump(
                    data,
                    default_flow_style=False,
                    encoding='utf-8',
                    allow_unicode=True
                ).decode('utf-8')
            )

class TFAConfig(BaseConfig):
    """
    Class to handle the TFA yaml file which contains secrets for e.g. TOTP
    Config file is located at /etc/ajenti/tfa.yml and should have the following
    structure :
    totp:
      user@auth_id:
        secret_id:
          created: DATE
          description: DESCRIPTION
          secret: random key in base32 with 32 chars
    """

    def __init__(self):
        BaseConfig.__init__(self)
        self.data = {}
        self.path = '/etc/ajenti/tfa.yml'
        self.verify_totp = {}

    def ensure_structure(self):
        self.data.setdefault('users', {})

    def get_user_totp_secrets(self, userid):
        with open(self.path, 'r') as tfa:
            tfa_config = yaml.load(tfa, Loader=yaml.SafeLoader).get('users', {})
            user_secrets = tfa_config.get(userid, {}).get('totp', [])
        return [details['secret'] for details in user_secrets]

    def append_user_totp(self, data):
        config = self._read()
        userid = data['userid']
        if config['users'].get(userid, {}).get('totp', []):
            config['users'][userid]['totp'].append(data['secret_details'])
            self.verify_totp[userid] = None
        else:
            config['users'][userid] = {'totp': [data['secret_details']]}
        self._save(config)
        self.load()

    def delete_user_totp(self, data):
        config = self._read()
        userid = data['userid']
        totps = config['users'].get(userid, {}).get('totp', [])
        for secret in totps:
            if str(secret['created']) == data['timestamp']:
                if len(totps) == 1:
                    # Remove completely user entry
                    del config['users'][userid]
                else:
                    config['users'][userid]['totp'].remove(secret)
                break
        self._save(config)
        self.load()

    def _read(self):
        if os.path.exists(self.path):
            with open(self.path, 'r') as tfa:
                return yaml.load(tfa, Loader=yaml.SafeLoader)
        else:
            return {'users': {}}

    def load(self):
        if os.path.exists(self.path):
            os.chmod(self.path, 384)  # 0o600
            with open(self.path, 'r') as tfa:
                self.data = yaml.load(tfa, Loader=yaml.SafeLoader).get('users', {})
            # Don't keep secrets in memory and prepare verify values per user involved
            for userid, tfa_methods in self.data.items():
                self.verify_totp[userid] = None
                for tfa_method, values in tfa_methods.items():
                    for entry in values:
                        entry['secret'] = ''
        else:
            self.ensure_structure()

    def _save(self, data):
        with open(self.path, 'w') as tfa:
            tfa.write(
                yaml.safe_dump(
                    data,
                    default_flow_style=False,
                    encoding='utf-8',
                    allow_unicode=True
                ).decode('utf-8')
           )
        os.chmod(self.path, 384)  # 0o600

class AjentiUsers(BaseConfig):
    """
    Class to handle the users config file for the auth-user plugin.
    Config file is located at /etc/ajenti/users.yml and should have the following
    structure :
    users:
      username:
        email: ...@...
        password: hash
        permissions: {}
        uid: int
        fs_root: file system root directory
    """

    def __init__(self, path):
        BaseConfig.__init__(self)
        self.data = None
        self.path = os.path.abspath(path)

    def __str__(self):
        return self.path

    def load(self):
        # Find default users file
        if not self.path:
            # Check for users file in /etc/ajenti/users.yml
            if os.path.isfile('/etc/ajenti/users.yml'):
                config_path = '/etc/ajenti/users.yml'
            elif os.path.isfile(os.path.join(sys.path[0], 'users.yml')):
                # Try local users file
                config_path = os.path.join(sys.path[0], 'users.yml')

        if not os.path.exists(self.path):
            logging.error(f'Users file "{self.path}" not found')
            self.data = {'users': {}}
        else:
            if os.geteuid() == 0:
                os.chmod(self.path, 384)  # 0o600
            with open(self.path, 'r') as users:
                self.data = yaml.load(users, Loader=yaml.SafeLoader)
            if self.data['users'] is None:
                self.data['users'] = {}

    def save(self):
        with open(self.path, 'w') as f:
            f.write(yaml.safe_dump(self.data, default_flow_style=False, encoding='utf-8', allow_unicode=True).decode('utf-8'))

@interface
class UserConfigProvider():
    id = None
    name = None

    def __init__(self, context):
        self.data = None

    def load(self):
        raise NotImplementedError

    def harden(self):
        raise NotImplementedError

    def save(self):
        raise NotImplementedError

class UserConfigError(Exception):
    def __init__(self, message):
        self.message = message

    def __str__(self):
        return self.message

@public
@service
class UserConfigService():
    def __init__(self, context):
        self.context = context

    def get_provider(self):
        provider_id = aj.config.data['auth'].get('provider', 'os')
        for provider in UserConfigProvider.all(self.context):
            if provider.id == provider_id:
                return provider
        raise UserConfigError(f'User config provider {provider_id} is unavailable')

@component(UserConfigProvider)
class UserConfig(UserConfigProvider):
    id = 'os'
    name = 'OS users'

    def __init__(self, context):
        UserConfigProvider.__init__(self, context)
        username = pwd.getpwuid(os.getuid())[0]
        _dir = os.path.expanduser(f'~{username}/.config')
        if not os.path.exists(_dir):
            os.makedirs(_dir)
        self.path = os.path.join(_dir, 'ajenti.yml')
        if os.path.exists(self.path):
            self.load()
        else:
            self.data = {}

    def load(self):
        self.data = yaml.load(open(self.path), Loader=yaml.SafeLoader)

    def harden(self):
        os.chmod(self.path, stat.S_IRWXU)

    def save(self):
        with open(self.path, 'w') as f:
            f.write(yaml.safe_dump(
                self.data, default_flow_style=False, encoding='utf-8', allow_unicode=True
            ).decode('utf-8'))
        self.harden()
